18-websockets/classworks/artyom_chernov/client/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,77 @@
+import React, { useEffect, useState } from 'react';
+import Message from './components/Message';
+import './Chat.css';
+
+import socket from '../../socket.connection';
+
+function Chat() {
+  const [userName, setUserName] = useState("noname");
+  const [messageValue, setMessageValue] = useState('');
+  const [messages, setMessages] = useState([]);
+  const [messageWriter, setMessageWriter] = useState('');
+
+  useEffect(() => {
+    let id;
+
+    socket.on("new-message-received", data => {
+      setMessages([...data])
+    })
+
+    socket.on("new-message-typing-stop", data => {
+      setMessageWriter(data.userName)
+      if(id) {
+        clearTimeout(id)
+      }
+      id = setTimeout(() => {
+        setMessageWriter('')
+      }, 3000);
+    })
+  }, []);
+
+  const sendMessage = () => {
+    setMessageValue('')
+    socket.emit('new-message', { userName, messageValue })
+  }
+
+  const onMessageChange = (e) => {
+    socket.emit('new-message-typing', { userName });
```

Comment:
if the user is a super typer man, you will get a huge number of requests from him. A good practice is to use the debounce method for calling function no more than one time per 2-4 secs.
Unfortunately, javascript doesn't have this method. You can write it by yourself or use some library.